### PR TITLE
Add automation steps for creating and converting apple push notification cert

### PR DESCRIPTION
**Pull request by metanerd**
#### Summary

Because this task needs to be performed on a yearly occurrence for multiple apps, it makes sense to ease this process. 
A future feature will hopefully also cover the manual upload and download tasks with the apple portal.

### CODE REVIEW

**Comment by jupenur** [comment]
@metanerd I suppose you meant to add me and @agnivade as reviewers, not assignees? Just making sure.

**Comment by metanerd** [comment]
> @metanerd I suppose you meant to add me and @agnivade as reviewers, not assignees? Just making sure.

Totally, sorry about that.

File: renew_apple_cert/create_csr.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		return nil, err
+	}
+	marshaledKey := x509.MarshalPKCS1PrivateKey(key)
+	pemPrivateKey := pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: marshaledKey,
+		},
+	)
+	err = ioutil.WriteFile(dirCsr+"/"+app+".key", pemPrivateKey, 0644)
```
Comment (jupenur) [inline renew_apple_cert/create_csr.go:95]: In case this is ever run in a multi-user environment: surely the private key and everything related to it should be `0600` max? That applies to parent directories, conversion output etc.

Comment (metanerd) [reply renew_apple_cert/create_csr.go]: Agreed.

File: renew_apple_cert/convert_certificate.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+func main() {
+	err := godotenv.Load()
+	if err != nil {
+		log.Fatal("Error loading .env file")
+	}
+
+	dirCsr := os.Getenv("DIR_CSR")
+	dirDownloaded := os.Getenv("DIR_DOWNLOADED")
+	dirConverted := os.Getenv("DIR_CONVERTED")
```
Comment (agnivade) [inline renew_apple_cert/convert_certificate.go:32]: I think instead of 3 separate directories, we can just have one single directory path, and let the program create the csr, downloaded, and converted sub-directories under it. That way it's more consistent and organized.

File: renew_apple_cert/create_csr.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	template := x509.CertificateRequest{
+		Subject:            subj,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+	}
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, &template, key)
+	if err != nil {
+		return err
+	}
+	cr := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrBytes})
+	err = ioutil.WriteFile(dirCsr+"/"+i.app+".csr", cr, fileMode)
```
Comment (agnivade) [inline renew_apple_cert/create_csr.go:131]: path.Join

File: renew_apple_cert/create_csr.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	province       string
+	locality       string
+	organization   string
+	email          string
+}
+
+const fileMode = 0600
+
+func newInputCSR(app string, applePushTopic string, country string, province string, locality string, organization string, email string) (i *inputCSR) {
+	i = &inputCSR{
```
Comment (agnivade) [inline renew_apple_cert/create_csr.go:30]: Avoid named returns. Just `return &inputCSR{..}`

**Comment by metanerd** [comment]
I am taking suggestions on how to approach testing the conversion of the cert, because I need the key and the cert for this, which are both secrets and then they will also expire. Maybe the verify at the end is enough for now?

**Comment by agnivade** [comment]
Can we generate a key and cert during a test?

And I think my suggestion was somehow misunderstood. I was suggesting to create two separate main packages under renew_apple_cert. Like:
cmd/renew_apple_cert/create_csr/main.go
cmd/renew_apple_cert/convert_certification/main.go

I don't see any main packages in the latest change. How are we planning to run it?

There are also weird usages of init functions, which are not necessary.

Also, wondering your thoughts on
https://github.com/mattermost/mattermost-push-proxy/pull/73#discussion_r443687802
https://github.com/mattermost/mattermost-push-proxy/pull/73#discussion_r443710551

In short, this PR needs some work. Sorry if all of this sounds laborious. I am happy to do it myself if you prefer.

File: renew_apple_cert/.env.example [hunk 5]
```diff
@@ -0,0 +1,11 @@
+APP=mattermost
```
[the remark this answers is not in the document]
Comment (metanerd) [reply renew_apple_cert/.env.example]: I heard in the dev meeting, I think, that now godotenv is considered to be used for the config files in our other projects too. Maybe I misheard.

File: renew_apple_cert/convert_certificate.go (hunk 2 above)
Comment (metanerd) [reply renew_apple_cert/convert_certificate.go]: Agreed.

**Comment by metanerd** [comment]
> In short, this PR needs some work. Sorry if all of this sounds laborious. I am happy to do it myself if you prefer.
I have to touch it again soon anyway to renew some additional certs and then also integrate with the apple dev api. I hope to address your items of discussion then. Let me convert it to a draft for now and then remove you guys from review.

File: renew_apple_cert/.env.example (hunk 5 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply renew_apple_cert/.env.example]: Can you give me a link to that? I must have missed it.

[the remark this answers is not in the document]
Comment (metanerd) [reply renew_apple_cert/.env.example]: https://community.mattermost.com/core/pl/94u8wezfbf8kukrqqymqpurpaa I think it is in the recording of this dev meeting, but not really certain when exactly this was raised.

**Comment by agnivade** [comment]
@metanerd - I have improved the code now. Please give it a try locally with all the steps and let me know if it works. The primary external thing that has changed is that now the tool takes a config.json file instead of dot env files.

**Comment by agnivade** [comment]
ping @metanerd.

**Comment by metanerd** [comment]
@agnivade Great work! Thank you for the revamp! I am going to merge it now!

**Comment by metanerd** [comment]
https://github.com/mattermost/mattermost-push-proxy/pull/81 I forgot some README fixes. I will also update the documentation.
